grails-app/services/com/unifina/service/StreamService.groovy

```diff
@@ -120,41 +120,43 @@ class StreamService {
 
 	// Ref to Kafka will be abstracted out when Feed abstraction is reworked
 
-	void sendMessage(Stream stream, String partitionKey, long timestamp, String content, StreamMessage.ContentType contentType,
-					 StreamMessage.SignatureType signatureType = StreamMessage.SignatureType.SIGNATURE_TYPE_NONE, String address = "", String signature = null) {
-		int streamPartition = partitioner.partition(stream, partitionKey)
-		StreamMessage msg = new StreamMessageV30(stream.id, streamPartition, timestamp,
-			0, address, (Long) null, 0, contentType, content, signatureType, signature)
-
-		String kafkaPartitionKey = "${stream.id}-$streamPartition"
+	void sendMessage(StreamMessage msg) {
+		String kafkaPartitionKey = "${msg.getStreamId()}-${msg.getStreamPartition()}"
 		kafkaService.sendMessage(msg, kafkaPartitionKey)
 	}
 
 	@CompileStatic
 	void sendMessage(Stream stream, Map message) {
```

Comment:
I meant that we could keep only `sendMessage(StreamMessage msg)` above and delete all the other variants of `sendMessage`. Is that doable? That way we could roll the responsibility for the message chaining and sequence numbering to the caller, which is where it should be.

Reply:
I guess it is doable but it will involve a lot a code duplication in many files to basically do the same thing as these functions do. The `gson` object has to be duplicated in every file to convert to JSON strings. I tried and ran into some issue because the `com.google.Gson` class is not serializable and apparently the modules need to be serializable by `ModuleTestHelper.Builder`.